beacon_node/operation_pool/src/lib.rs

```diff
@@ -215,28 +218,128 @@ impl<T: EthSpec> OperationPool<T> {
-                AttMaxCover::new(att, state, reward_cache, total_active_balance, spec)
-            })
+    ) -> Vec<(&CompactAttestationData, CompactIndexedAttestation<T>)> {
+        let mut cliqued_atts: Vec<(&CompactAttestationData, CompactIndexedAttestation<T>)> = vec![];
+
+        if let Some(AttestationDataMap {
+            aggregate_attestations,
+            unaggregate_attestations,
+        }) = all_attestations.get_attestation_map(checkpoint_key)
+        {
+            for (data, aggregates) in aggregate_attestations {
+                if data.slot + spec.min_attestation_inclusion_delay <= state.slot()
+                    && state.slot() <= data.slot + T::slots_per_epoch()
+                {
+                    let aggregates: Vec<&CompactIndexedAttestation<T>> = aggregates
+                        .iter()
+                        .filter(|indexed| {
+                            validity_filter(&AttestationRef {
+                                checkpoint: checkpoint_key,
+                                data: &data,
+                                indexed,
+                            })
+                        })
+                        .collect();
+                    *num_valid += aggregates.len() as i64;
+
+                    // derive cliques for current attestation data
+                    let cliques = bron_kerbosch(&aggregates, is_compatible);
+
+                    // aggregate each cliques corresponding attestaiions
+                    let mut clique_aggregates: Vec<CompactIndexedAttestation<_>> = cliques
+                        .iter()
+                        .map(|clique| {
+                            let mut res_att = aggregates[clique[0]].clone();
+                            for ind in clique.iter().skip(1) {
+                                res_att.aggregate(&aggregates[*ind]);
+                            }
+                            res_att
+                        })
+                        .collect();
+
+                    let mut indices_to_remove = Vec::new();
+                    clique_aggregates
+                        .sort_by(|a, b| a.attesting_indices.len().cmp(&b.attesting_indices.len()));
+                    for (index, clique) in clique_aggregates.iter().enumerate() {
+                        for bigger_clique in clique_aggregates.iter().skip(index + 1) {
+                            if clique
+                                .aggregation_bits
+                                .is_subset(&bigger_clique.aggregation_bits)
+                            {
+                                indices_to_remove.push(index);
+                                break;
+                            }
+                        }
+                    }
+                    for index in indices_to_remove.iter().rev() {
+                        clique_aggregates.remove(*index);
+                    }
+                    let mut clique_aggregates = clique_aggregates.into_iter();
```

Comment:
This line contains the bug!! It _moves_ `clique_aggregates` into an iterator, which means that later when we iterate by mutable reference (`for mut clique_aggregate in &mut clique_aggregates { .. }`) we actually _consume_ each attestation that we've iterated. That means whenever there are unaggregated attestations to add we throw away a large portion of our attestations! This also explains why nodes running with `--subscribe-all-subnets` are most affected!

I removed this line in favour of just iterating over the vec (which _won't_ consume the elements) and can confirm that it fixes the block profitability :tada:

Reply:
This is the patch I applied:

```patch
diff --git a/beacon_node/operation_pool/src/lib.rs b/beacon_node/operation_pool/src/lib.rs
index 3daf1cd8b..e45e3e2e7 100644
--- a/beacon_node/operation_pool/src/lib.rs
+++ b/beacon_node/operation_pool/src/lib.rs
@@ -268,7 +268,6 @@ impl<T: EthSpec> OperationPool<T> {
                             res_att
                         })
                         .collect();
-
                     let mut indices_to_remove = Vec::new();
                     clique_aggregates
                         .sort_by(|a, b| a.attesting_indices.len().cmp(&b.attesting_indices.len()));
@@ -283,10 +282,10 @@ impl<T: EthSpec> OperationPool<T> {
                             }
                         }
                     }
+
                     for index in indices_to_remove.iter().rev() {
                         clique_aggregates.remove(*index);
                     }
-                    let mut clique_aggregates = clique_aggregates.into_iter();
 
                     // aggregate unaggregate attestations into the clique aggregates
                     // if compatible
@@ -299,7 +298,7 @@ impl<T: EthSpec> OperationPool<T> {
                             })
                         }) {
                             *num_valid += 1;
-                            for mut clique_aggregate in &mut clique_aggregates {
+                            for clique_aggregate in &mut clique_aggregates {
                                 if !clique_aggregate
                                     .attesting_indices
                                     .contains(&attestation.attesting_indices[0])
@@ -310,7 +309,8 @@ impl<T: EthSpec> OperationPool<T> {
                         }
                     }
 
-                    cliqued_atts.extend(clique_aggregates.map(|indexed| (data, indexed)));
+                    cliqued_atts
+                        .extend(clique_aggregates.into_iter().map(|indexed| (data, indexed)));
                 }
             }
             // include aggregated attestations from unaggregated attestations whose

```